_source/_posts/2022-06-01-jhipster-digital-ocean.md

```diff
@@ -0,0 +1,419 @@
+```
+After creating a cluster, `doctl` adds a configuration context to kubectl and makes it active, so you can start monitoring your cluster right away with k9s. First, apply the resources configuration to the DigitalOcean cluster:
+
+```bash
+./kubectl-apply.sh -f
+```
+
+Monitor the deployment with k9s:
+
+```bash
+k9s -n demo
+```
+
+{% img blog/jhipster-digital-ocean/k9s-do-cluster.png alt:"k9s user interface monitoring DigitalOcean Kubernetes" width:"800" %}{: .center-image }
+
+Once you see the `jhipster-registry` pods are up, set up port forwarding again so you can also monitor the status of the services in the registry UI.
+
+### Increase CPU
+
+The three-node cluster might not provide enough CPU for this microservices architecture, so you might see `insufficient CPU` logs for some pods. You can get the pod events with `kubectl describe`:
+
+```bash
+kubectl describe pod jhipster-registry-0 -n demo
+```
+```text
+Events:
+  Type     Reason                  Age                   From                     Message
+  ----     ------                  ----                  ----                     -------
+  Normal   NotTriggerScaleUp       4m30s                 cluster-autoscaler       pod didn't trigger scale-up:
+  Warning  FailedScheduling        69s (x3 over 95s)     default-scheduler        0/3 nodes are available: 3 Insufficient cpu.
+```
+Increase the number of nodes with `doctl`:
+
+```bash
+doctl k cluster node-pool update do1 do1-default-pool --count 4
+```
+
+### Increase Storage
+
+For the `store-mongodb` stateful set with 3 replicas, some of the pods did not run due to _unbound immediate PersistentVolumeClaims_. When inspecting the pod events, I found out there was a failure in volume provisioning:
+
+```text
+Events:
+  Type     Reason                Age                  From                                                                       Message
+  ----     ------                ----                 ----                                                                       -------
+  Normal   Provisioning          3m6s (x10 over 11m)  dobs.csi.digitalocean.com_master-do3_cebb3451-5eba-4805-956c-753ec148e2ea  External provisioner is provisioning volume for claim "demo/datadir-store-mongodb-2"
+  Warning  ProvisioningFailed    3m5s (x10 over 11m)  dobs.csi.digitalocean.com_master-do3_cebb3451-5eba-4805-956c-753ec148e2ea  failed to provision volume with StorageClass "do-block-storage": rpc error: code = ResourceExhausted desc = volume limit (10) has been reached. Current number of volumes: 10. Please contact support.
+  Normal   ExternalProvisioning  103s (x43 over 11m)  persistentvolume-controller                                                waiting for a volume to be created, either by external provisioner "dobs.csi.digitalocean.com" or manually created by system administrator
+```
+
+As instructed by the event message, I contacted DigitalOcean support and they fixed it.
```

Comment:
I created a ticket with DigitalOcean Support to try and get this fixed.

Reply:
Support replied with the following:

> Upon checking your account I can see there is k8s cluster with name 'do1' and all the PV's are bound successfully to the cluster:
> ...
> As far I see the pod "store-mongodb-1" is in pending state because there is insufficient cpu on all the nodes for the pod to get deployed:
> ```
> % k get events -A
> NAMESPACE   LAST SEEN   TYPE      REASON              OBJECT                MESSAGE
> demo        97s         Warning   FailedScheduling    pod/store-mongodb-1   0/3 nodes are available: 3 Insufficient CPU.
> demo        3m50s       Normal    NotTriggerScaleUp   pod/store-mongodb-1   pod didn't trigger scale-up:
> ```
> I would suggest to increase the nodes in the cluster so that the pod can deploy without any problem. 

I tried the above command:

```
doctl k cluster node-pool update do1 do1-default-pool --count 4
```

And then restarted all my pods:

```
kubectl -n demo rollout restart deploy
```

After waiting several minutes, it doesn't seem like everything is starting.

<img width="1442" alt="Screen Shot 2022-05-19 at 12 17 53" src="https://user-images.githubusercontent.com/17892/169291447-86f21c77-6839-418f-ac9e-a967df0d51e3.png">

Reply:
I provided this information to Digital Ocean support. Here's what they said:

On describing your one of the pod 'store-8fb4df89f-b5mz4' I could see that you have configured resource limit. To isolate the issue can you please modify the value of resource limit or if possible please set the minimal required value for resource limit and let see issue is fixed or not! Please refer to the below link for more information.  https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/  If issue persist, Will it possible for you to use larger node.  The small node sizes are not intended for production workloads. These node sizes are there for experimentation and educational purposes. After taking into account the resource usage of the containerized infrastructure there is not a lot or resources left for production workloads to run in a stable or sustainable manner. 

Using many small nodes in your cluster can be an inefficient use of compute resources. Customers actually get more resources per dollar by using moderately sized nodes as each new node requires overhead of components such as cilium, DO node agents, and CSI containers. We recommend for customers to have at least 3 moderately sized nodes as opposed to say 6 very small nodes. This way the overhead of infrastructure takes less of your resources and leaves more for you to use. That being said customers should still maintain multiple nodes to increase the fault tolerance of their deployments.

Reply:
4 nodes of size s-2vcpu-4gb-intel worked for me to solve "insufficient cpu" errors. After cpu was enough, I had the "unbound immediate PersistentVolumeClaims" error, preventing mongodb containers to start. I listed persistent volume claims `kubectl get pvc -n demo`, and then I did `kubectl describe pvc datadir-store-mongodb-2 -n demo` and saw the provisioning error in the events. I created the support ticket and they fixed it, they responded:

"I understand that your pods not getting created due to volume provision issues. Upon checking your cluster  I could see your volume limit was 10 and you already have 10 volumes in your account. I have gone ahead and increased your volume limit please check your pod, it should be created now."

Reply:
I'll try it over from the beginning and see what I find.